File: adapter/src/main/kotlin/io/goooler/demoapp/adapter/rv/core/BindingAdapters.java
package io.goooler.demoapp.adapter.rv.core;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.BindingAdapter;
import androidx.recyclerview.widget.RecyclerView;

import java.util.Collections;
import java.util.List;

public interface BindingAdapters {

  @BindingAdapter("binding_rv_dataList")
  static <M extends IVhModelType> void bindingSetList(
    @NonNull RecyclerView recyclerView,
    @Nullable List<M> list
  ) {
    if (recyclerView.getAdapter() instanceof IMutableRvAdapter) {
      if (list == null) {
        list = Collections.emptyList();
      }
      //noinspection unchecked
      ((IMutableRvAdapter<M>) recyclerView.getAdapter()).setList(list);
    }
  }

  @BindingAdapter("binding_rv_refreshItems")
  static <M extends IVhModelType> void bindingRefreshItems(
    @NonNull RecyclerView recyclerView,
    @Nullable List<M> list
  ) {
    if (recyclerView.getAdapter() instanceof IMutableRvAdapter) {
      if (list == null) {
        list = Collections.emptyList();
      }
      //noinspection unchecked
      ((IMutableRvAdapter<M>) recyclerView.getAdapter()).refreshItems(list);
    }
  }
}

File: adapter/rv/src/main/kotlin/io/github/goooler/adapter/rv/internal/BindingAdapters.kt
package io.github.goooler.adapter.rv.internal

import androidx.databinding.BindingAdapter
import androidx.recyclerview.widget.RecyclerView
import io.github.goooler.adapter.rv.core.IVhModelType

/**
 * Used for generating binding adapters, mark the class as internal to avoid using in Kotlin.
 */
internal object BindingAdapters {
  @BindingAdapter("binding_rv_dataList")
  @JvmStatic
  fun <M : IVhModelType> RecyclerView.bindingSetList(list: List<M>?) {
    @Suppress("UNCHECKED_CAST")
    (adapter as? IMutableRvAdapter<M>)?.list = list.orEmpty()
  }

  @BindingAdapter("binding_rv_refreshItems")
  @JvmStatic
  fun <M : IVhModelType> RecyclerView.bindingRefreshItems(vararg items: Triple<Int, M, Any?>) {
    @Suppress("UNCHECKED_CAST")
    (adapter as? IMutableRvAdapter<M>)?.refreshItems(*items)
  }
}
